In [1]:
import pandas as pd
import numpy as np
#this are important as I start with project i usually import this libraries because pandas is use to read files and data
#numpy is used to help in mathametical operations

In [2]:
import mysql.connector

# Create connection
conn = mysql.connector.connect(
    host="localhost",
    user="root",        # your username
    password="Mukka@7979",
    database="raodo"
)

print("✅ Connected to MySQL")

✅ Connected to MySQL


In [3]:
customers = pd.read_sql("select * from customers", conn)

C:\Users\surya\AppData\Local\Temp\ipykernel_27668\4208339579.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  customers = pd.read_sql("select * from customers", conn)


In [4]:
customers.head()

,customer_id,company_name,industry,company_size,country_code,country_name,timezone,contact_email,contact_name,signup_date,signup_source,is_active,churned_at,churn_reason,nps_score,created_at,updated_at
0,1,Rush Platform,E-commerce,small,FR,France,Europe/Paris,michael.baker@corp.net,Michael Baker,2023-01-31,partner,0,2024-02-26 10:13:36,Outgrew the platform,NaN,2026-03-21 20:59:37,2026-03-21 20:59:37
1,2,TrekStack,Manufacturing,small,CA,Canada,America/Toronto,emily.moore+work@startup.co,Emily Moore,2023-06-09,referral,1,NaT,None,NaN,2026-03-21 20:59:37,2026-03-21 20:59:37
2,3,Wave Group,Education,enterprise,US,United States,America/New_York,olga.robinson@wavegroup.com,Olga Robinson,2023-10-28,organic,1,NaT,None,NaN,2026-03-21 20:59:37,2026-03-21 20:59:37
3,4,KeenSoftware,Technology,large,GB,United Kingdom,Europe/London,lars.jackson@outlook.com,Lars Jackson,2024-04-09,partner,1,NaT,None,7.0,2026-03-21 20:59:37,2026-03-21 20:59:37
4,5,QuestAnalytics,Logistics,large,US,United States,America/Chicago,joshua.green@yahoo.com,Joshua Green,2023-10-04,partner,1,NaT,None,NaN,2026-03-21 20:59:37,2026-03-21 20:59:37


In [5]:
billing_invoices = pd.read_sql("select * from billing_invoices", conn)

C:\Users\surya\AppData\Local\Temp\ipykernel_27668\4089747319.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  billing_invoices = pd.read_sql("select * from billing_invoices", conn)


In [6]:
billing_invoices.head()

,invoice_id,customer_id,subscription_id,invoice_date,due_date,amount_usd,tax_usd,total_usd,status,payment_method,currency,paid_at,created_at
0,1,963,1716,2025-08-06,2025-09-05,245.44,28.67,274.11,paid,credit_card,USD,2025-08-26 06:47:06,2026-03-21 21:11:03
1,2,1084,1707,2024-10-19,2024-11-18,99.76,7.04,106.80,paid,wire,USD,2024-10-23 02:44:11,2026-03-21 21:11:03
2,3,381,1747,2025-05-20,2025-06-19,117.82,3.43,121.25,paid,credit_card,USD,2025-05-25 00:34:12,2026-03-21 21:11:03
3,4,214,120,2024-02-20,2024-03-21,90.00,1.57,91.57,overdue,bank_transfer,GBP,NaT,2026-03-21 21:11:03
4,5,449,1150,2024-11-07,2024-12-07,67.54,1.58,69.12,void,wire,INR,NaT,2026-03-21 21:11:03


In [7]:
feature_flags = pd.read_sql("select * from feature_flags", conn)

C:\Users\surya\AppData\Local\Temp\ipykernel_27668\4074788208.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  feature_flags = pd.read_sql("select * from feature_flags", conn)


In [8]:
feature_flags.head()

,flag_id,plan_id,feature_key,feature_name,is_enabled,enabled_at,created_at
0,1,1,project_boards,Core project management boards,1,2023-01-25 18:02:58,2026-03-21 21:14:55
1,2,1,ai_task_suggest,AI-powered task suggestions,0,NaT,2026-03-21 21:14:55
2,3,1,gantt_charts,Gantt chart visualization,0,NaT,2026-03-21 21:14:55
3,4,1,time_tracking,Built-in time tracking,0,NaT,2026-03-21 21:14:55
4,5,1,resource_mgmt,Resource management & allocation,0,NaT,2026-03-21 21:14:55


In [9]:
plans = pd.read_sql("select * from plans", conn)

C:\Users\surya\AppData\Local\Temp\ipykernel_27668\1837956911.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  plans = pd.read_sql("select * from plans", conn)


In [10]:
plans.head()

,plan_id,plan_name,plan_tier,monthly_price_usd,annual_price_usd,max_users,max_projects,storage_gb,is_active,created_at
0,1,Free,free,0.0,0.0,5.0,3.0,1,1,2026-03-21 20:59:37
1,2,Starter,starter,12.0,120.0,15.0,10.0,5,1,2026-03-21 20:59:37
2,3,Starter Plus,starter,19.0,190.0,25.0,20.0,10,1,2026-03-21 20:59:37
3,4,Professional,professional,39.0,390.0,50.0,50.0,25,1,2026-03-21 20:59:37
4,5,Professional Plus,professional,59.0,590.0,100.0,100.0,50,1,2026-03-21 20:59:37


In [11]:
subscriptions = pd.read_sql("select * from subscriptions", conn)

C:\Users\surya\AppData\Local\Temp\ipykernel_27668\3044609939.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  subscriptions = pd.read_sql("select * from subscriptions", conn)


In [12]:
subscriptions.head()

,subscription_id,customer_id,plan_id,status,billing_cycle,start_date,end_date,mrr_usd,discount_pct,trial_end_date,cancellation_reason,created_at
0,1,1,1,cancelled,annual,2023-01-31,2023-04-20,0.00,15.0,None,Missing features,2026-03-21 21:10:51
1,2,1,5,cancelled,annual,2023-04-20,2024-02-26,44.25,10.0,None,Switched competitor,2026-03-21 21:10:51
2,3,2,1,paused,monthly,2023-06-09,None,0.00,0.0,None,None,2026-03-21 21:10:51
3,4,3,1,active,annual,2023-10-28,None,0.00,20.0,None,None,2026-03-21 21:10:51
4,5,4,1,expired,annual,2024-04-09,2024-08-10,0.00,15.0,None,Budget cuts,2026-03-21 21:10:51


In [13]:
support_tickets = pd.read_sql("select * from support_tickets", conn)

C:\Users\surya\AppData\Local\Temp\ipykernel_27668\2626980804.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  support_tickets = pd.read_sql("select * from support_tickets", conn)


In [14]:
support_tickets.head()

,ticket_id,customer_id,member_id,category,priority,status,subject,created_at,first_response_at,resolved_at,satisfaction_score,escalated,tags
0,1,518,6603.0,data_export,high,open,Account merge request,2025-05-19 03:43:53,NaT,NaT,NaN,1,None
1,2,810,NaN,technical,medium,closed,Custom workflow automation broken,2025-09-22 07:22:01,2025-09-25 00:22:01,2025-09-27 07:22:01,4.0,0,"training,onboarding,gdpr"
2,3,1080,NaN,onboarding,high,in_progress,Need help with API authentication,2025-10-29 11:11:29,2025-10-29 21:11:29,NaT,NaN,0,integration
3,4,1009,NaN,technical,medium,open,Time tracking sync issues,2025-02-03 09:09:03,NaT,NaT,NaN,0,"onboarding,bug,security,integration"
4,5,840,NaN,technical,medium,waiting_on_customer,SSO configuration help needed,2024-04-03 13:09:04,2024-04-03 20:09:04,NaT,NaN,0,None


In [15]:
team_members = pd.read_sql("select * from team_members", conn)

C:\Users\surya\AppData\Local\Temp\ipykernel_27668\2909548890.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  team_members = pd.read_sql("select * from team_members", conn)


In [16]:
team_members.head()

,member_id,customer_id,email,full_name,role,is_active,invited_at,joined_at,last_login_at,created_at
0,1,1,joseph.nelson@rushplatform.com,Joseph Nelson,admin,1,2023-03-28 06:12:21,2023-04-06 21:02:07,2024-02-04 08:29:58,2026-03-21 20:59:37
1,2,1,sofia.wright@rushplatform.com,Sofia Wright,admin,1,2023-03-25 14:38:48,2023-04-02 05:26:38,2024-01-03 06:09:59,2026-03-21 20:59:37
2,3,1,yuki.martinez@rushplatform.com,Yuki Martinez,member,1,2023-03-09 00:47:39,2023-03-10 05:59:34,2023-05-08 15:58:07,2026-03-21 20:59:37
3,4,2,jennifer.adams@trekstack.com,Jennifer Adams,admin,1,2023-07-26 07:28:51,2023-08-02 06:06:53,2025-09-26 13:59:24,2026-03-21 20:59:37
4,5,2,richard.williams@trekstack.com,Richard Williams,billing,0,2023-07-09 22:14:41,NaT,NaT,2026-03-21 20:59:37


In [17]:
from pymongo import MongoClient

# Connect to MongoDB
client = MongoClient("mongodb://localhost:27017/")
db = client["nimbus_events"]

collection = db["user_activity_logs"]

data = list(collection.find())

df_mongo = pd.DataFrame(data)

print("Shape:", df_mongo.shape)
df_mongo.head()

Shape: (51485, 17)


,_id,customer_id,member_id,timestamp,event_type,page_url,session_duration_sec,browser,os,device_type,timezone,ip_address,userId,customerId,referrer,feature,userID
0,69bed815ec39b1e38f11402a,229,410.0,2024-01-16 10:13:48,form_submit,/projects/1779,2263.0,Chrome,iOS 17,mobile,UTC,60.214.112.115,NaN,NaN,NaN,NaN,NaN
1,69bed815ec39b1e38f11402b,14,2616.0,06/12/2024 13:27:31,search,/projects/3627,NaN,Safari,Windows 10,desktop,Asia/Kolkata,NaN,NaN,NaN,NaN,NaN,NaN
2,69bed815ec39b1e38f11402c,NaN,NaN,2025-09-03 13:23:25,notification_click,/api-docs,2915.0,NaN,NaN,NaN,America/Los_Angeles,NaN,1292.0,776,NaN,NaN,NaN
3,69bed815ec39b1e38f11402d,334,6066.0,2024-02-19 17:39:16,dashboard_view,/api-docs,730.0,NaN,NaN,NaN,America/Chicago,NaN,NaN,NaN,NaN,NaN,NaN
4,69bed815ec39b1e38f11402e,115,3753.0,2024-05-05T00:56:31.000+00:00,import,/team,2353.0,Safari Mobile,macOS 14,desktop,NaN,200.134.219.230,NaN,NaN,NaN,NaN,NaN


In [18]:
df_mongo.columns

Index(['_id', 'customer_id', 'member_id', 'timestamp', 'event_type',
       'page_url', 'session_duration_sec', 'browser', 'os', 'device_type',
       'timezone', 'ip_address', 'userId', 'customerId', 'referrer', 'feature',
       'userID'],
      dtype='object')

In [19]:
#next we have to follow is cleaning columns as below 
#as there is multiple user_id columns we have to combine them that which follows below 
#there are three columns which refer to user_id/member_id because there must be either customer or user so I combines all this three columns together
df_mongo['user_id'] = df_mongo['userId']
df_mongo['user_id'] = df_mongo['user_id'].combine_first(df_mongo['userID'])
df_mongo['user_id'] = df_mongo['user_id'].combine_first(df_mongo['member_id'])

In [20]:
df_mongo.shape

(51485, 18)

In [21]:
df_mongo[['customer_id', 'customerId']].info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 51485 entries, 0 to 51484
Data columns (total 2 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   customer_id  38576 non-null  object
 1   customerId   12909 non-null  object
dtypes: object(2)
memory usage: 804.6+ KB


In [22]:
# as we check there is customer_id also repeated so we have to combine them as well and create it into new column
df_mongo['customer_id_clean'] = df_mongo['customer_id']
df_mongo['customer_id_clean'] = df_mongo['customer_id_clean'].combine_first(df_mongo['customerId'])

In [23]:
df_mongo['customer_id_clean'] = pd.to_numeric(df_mongo['customer_id_clean'], errors='coerce')

In [24]:
df_mongo[['customer_id_clean']].head()

,customer_id_clean
0,229
1,14
2,776
3,334
4,115


In [25]:
# as we combine all customer id columns into one we don't need old two columns so, I'm removing them
df_mongo.drop(columns=['customer_id', 'customerId'], inplace=True)

In [26]:
#as we have to handle null values we have to check is there any null values are there

df_mongo['customer_id_clean'].isnull().sum()

np.int64(0)

In [27]:
df_mongo[['userId', 'userID', 'member_id']].info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 51485 entries, 0 to 51484
Data columns (total 3 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   userId     12909 non-null  float64
 1   userID     7750 non-null   float64
 2   member_id  30826 non-null  float64
dtypes: float64(3)
memory usage: 1.2 MB


In [28]:
df_mongo['user_id']

0          410.0
1         2616.0
2         1292.0
3         6066.0
4         3753.0
          ...   
51480    95410.0
51481     7071.0
51482     4482.0
51483     7573.0
51484     7119.0
Name: user_id, Length: 51485, dtype: float64

In [29]:
#now we delete duplicate user columns and change datatype into int

df_mongo.drop(columns=['userId', 'userID', 'member_id'], inplace=True)

In [30]:
df_mongo['user_id'] = df_mongo['user_id'].astype('Int64')

In [31]:
df_mongo['user_id'].isnull().sum()

np.int64(0)

In [32]:
#i don't like customer_id_clean so I transfer and delete the column with below query help
df_mongo['customer_id'] = df_mongo['customer_id_clean']
df_mongo.drop(columns=['customer_id_clean'], inplace=True)

In [33]:
df_mongo.columns

Index(['_id', 'timestamp', 'event_type', 'page_url', 'session_duration_sec',
       'browser', 'os', 'device_type', 'timezone', 'ip_address', 'referrer',
       'feature', 'user_id', 'customer_id'],
      dtype='object')

In [34]:
df_mongo['timestamp'].dtype
df_mongo['session_duration_sec'].dtype

dtype('float64')

In [35]:
# as they are already in the requried data types still I'm doing this because for my clarification
# timestamp
df_mongo['timestamp'] = pd.to_datetime(df_mongo['timestamp'], errors='coerce')

# session duration
df_mongo['session_duration_sec'] = pd.to_numeric(df_mongo['session_duration_sec'], errors='coerce')

C:\Users\surya\AppData\Local\Temp\ipykernel_27668\3965825287.py:3: FutureWarning: In a future version of pandas, parsing datetimes with mixed time zones will raise an error unless `utc=True`. Please specify `utc=True` to opt in to the new behaviour and silence this warning. To create a `Series` with mixed offsets and `object` dtype, please use `apply` and `datetime.datetime.strptime`
  df_mongo['timestamp'] = pd.to_datetime(df_mongo['timestamp'], errors='coerce')


In [36]:
#we are merging mongodb and sql data to do that we need to follow some check list like below
# should have same data type
df_mongo['customer_id'].dtype
customers['customer_id'].dtype

dtype('int64')

In [37]:
#as we need to merge with common column and below is the query for that
# and we have to follow inner join because we don't whether there is mismatch data and that would lead to wrong data
merged_df = df_mongo.merge(
    customers,
    on='customer_id',
    how='inner'
)


In [38]:
# to check whether merge is success or not
print("Merged shape:", merged_df.shape)
merged_df.head(15)

Merged shape: (50497, 30)


,_id,timestamp,event_type,page_url,session_duration_sec,browser,os,device_type,timezone_x,ip_address,...,contact_email,contact_name,signup_date,signup_source,is_active,churned_at,churn_reason,nps_score,created_at,updated_at
0,69bed815ec39b1e38f11402a,2024-01-16 10:13:48,form_submit,/projects/1779,2263.0,Chrome,iOS 17,mobile,UTC,60.214.112.115,...,ahmed.santos@linkdynamics.com,Ahmed Santos,2025-05-24,partner,1,NaT,None,NaN,2026-03-21 20:59:37,2026-03-21 20:59:37
1,69bed815ec39b1e38f11402b,2024-06-12 13:27:31,search,/projects/3627,NaN,Safari,Windows 10,desktop,Asia/Kolkata,NaN,...,hassan.clark@startup.co,Hassan Clark,2024-02-25,referral,1,NaT,None,NaN,2026-03-21 20:59:37,2026-03-21 20:59:37
2,69bed815ec39b1e38f11402c,2025-09-03 13:23:25,notification_click,/api-docs,2915.0,NaN,NaN,NaN,America/Los_Angeles,NaN,...,wei.rodriguez@arcsoftware.com,Wei Rodriguez,2023-08-19,referral,0,2024-05-27 03:08:50,Missing features,1.0,2026-03-21 20:59:37,2026-03-21 20:59:37
3,69bed815ec39b1e38f11402d,2024-02-19 17:39:16,dashboard_view,/api-docs,730.0,NaN,NaN,NaN,America/Chicago,NaN,...,donald.scott@corp.net,Donald Scott,2024-09-22,event,1,NaT,None,5.0,2026-03-21 20:59:37,2026-03-21 20:59:37
4,69bed815ec39b1e38f11402e,2024-05-05 00:56:31+00:00,import,/team,2353.0,Safari Mobile,macOS 14,desktop,NaN,200.134.219.230,...,david.lee@tech.dev,David Lee,2025-06-18,paid_ads,1,NaT,None,8.0,2026-03-21 20:59:37,2026-03-21 20:59:37
5,69bed815ec39b1e38f11402f,2023-03-15 04:37:32,file_upload,/projects,3274.0,Safari Mobile,macOS 15,mobile,America/New_York,151.5.58.175,...,raj.santos@yahoo.com,Raj Santos,2023-06-14,paid_ads,0,2025-09-17 08:42:34,Too expensive,NaN,2026-03-21 20:59:37,2026-03-21 20:59:37
6,69bed815ec39b1e38f114030,2023-09-03 16:32:54,share,/dashboard,NaN,Firefox,iOS 17,desktop,NaN,51.0.165.126,...,hassan.jones@outlook.com,Hassan Jones,2023-05-11,partner,1,NaT,None,3.0,2026-03-21 20:59:37,2026-03-21 20:59:37
7,69bed815ec39b1e38f114031,2025-05-31 08:35:59,feature_click,/calendar,NaN,Chrome Mobile,Linux,desktop,America/Chicago,NaN,...,michael.miller@xenonstack.com,Michael Miller,2023-10-16,event,0,2024-02-05 16:38:34,Switched to competitor,0.0,2026-03-21 20:59:37,2026-03-21 20:59:37
8,69bed815ec39b1e38f114032,2024-09-11 19:30:39,settings_change,/tasks,2149.0,Edge,Windows 10,desktop,America/Los_Angeles,NaN,...,kenji.garcia@company.io,Kenji Garcia,2025-03-15,event,1,NaT,None,NaN,2026-03-21 20:59:37,2026-03-21 20:59:37
9,69bed815ec39b1e38f114033,2024-05-28 07:16:01,feature_click,/settings,907.0,Chrome Mobile,Windows 11,tablet,America/Sao_Paulo,NaN,...,andrew.nelson@gmail.com,Andrew Nelson,2024-08-07,referral,0,2024-11-04 16:19:00,Missing features,NaN,2026-03-21 20:59:37,2026-03-21 20:59:37


In [39]:
#as we have timestamp values are null so, we are going to merge created_at and timestamp
merged_df['event_time'] = merged_df['timestamp'].combine_first(merged_df['created_at'])

In [40]:
merged_df.drop(columns=['timestamp'], inplace=True)

In [41]:
# to check whether there is null values or not
merged_df.isnull().sum()

_id                         0
event_type                  0
page_url                    0
session_duration_sec    10122
browser                  7626
os                       7626
device_type              7626
timezone_x               4571
ip_address              35344
referrer                40219
feature                 43272
user_id                     0
customer_id                 0
company_name                0
industry                   93
company_size               93
country_code               93
country_name               93
timezone_y                 93
contact_email              93
contact_name               93
signup_date                 0
signup_source              43
is_active                   0
churned_at              39525
churn_reason            41663
nps_score               20405
created_at                  0
updated_at                  0
event_time                  0
dtype: int64

In [42]:
#this isn't mandatory but for case of avg, some aggregate function
#we are replacing null values in session_duration to 0
merged_df['session_duration_sec'] = merged_df['session_duration_sec'].fillna(0)

In [43]:
# timezone as multiple columns so we are combine and delete remain columns for timezone and renamed as timezone
merged_df['timezone'] = merged_df['timezone_x'].combine_first(merged_df['timezone_y'])

merged_df.drop(columns=['timezone_x', 'timezone_y'], inplace=True)

In [44]:
# null handle is complete in my perspective
# because other columns like feature doesn't affected by null values
# now we gonna check duplicate values
merged_df.duplicated().sum()

np.int64(0)

In [45]:
print("Before cleaning:", df_mongo.shape)
print("After merge:", merged_df.shape)

Before cleaning: (51485, 14)
After merge: (50497, 29)


In [46]:
merged_df.isnull().sum()

_id                         0
event_type                  0
page_url                    0
session_duration_sec        0
browser                  7626
os                       7626
device_type              7626
ip_address              35344
referrer                40219
feature                 43272
user_id                     0
customer_id                 0
company_name                0
industry                   93
company_size               93
country_code               93
country_name               93
contact_email              93
contact_name               93
signup_date                 0
signup_source              43
is_active                   0
churned_at              39525
churn_reason            41663
nps_score               20405
created_at                  0
updated_at                  0
event_time                  0
timezone                    6
dtype: int64

So Far Handling null and duplicate values and change datatype is done so 90% of cleaning is done 
we don't have much time to analysis column by column so, we are using this way

Starting Hypothesis Testing H0 and H1

H0: Feature usage does not affect session duration
H1: Feature usage increases session duration

Using t-test, p-value = ___
Since p < 0.05 → reject H0

Conclusion: Feature usage leads to higher engagement

so, we have to find p value first on the basis of that we must choose to do that we are following below

In [47]:
from scipy.stats import ttest_ind

feature_users = merged_df[merged_df['feature'].notnull()]['session_duration_sec']
non_feature_users = merged_df[merged_df['feature'].isnull()]['session_duration_sec']

t_stat, p_value = ttest_ind(feature_users, non_feature_users, equal_var=False)

print("T-stat:", t_stat)
print("P-value:", p_value)

T-stat: 0.49600688094668105
P-value: 0.6199006532098695


Test Used:
Independent Two-Sample t-test

Result:
p-value = 0.6199

Decision:
Since p-value > 0.05, we fail to reject the null hypothesis

Conclusion:
There is no statistically significant difference in session duration between users who use features and those who do not

Q4: Engagment-based segmentation because I used to work as digital marketing and all the analysis are done based on this 
and i used to analysis the data and reduce lead cost based on this


In [48]:
#What is engagement segmentation?
#We group users/customers based on how active they are
# and as per that I'm counting total session duration per user 

In [49]:
#lets create engagment metric
engagement = merged_df.groupby('customer_id')['session_duration_sec'].sum().reset_index()

engagement.rename(columns={'session_duration_sec': 'total_session_time'}, inplace=True)

engagement.head()

,customer_id,total_session_time
0,1,62712.0
1,2,45816.0
2,3,54171.0
3,4,68474.0
4,5,61688.0


In [50]:
merged_df = merged_df.merge(engagement[['customer_id', 'segment']], on='customer_id', how='left')

KeyError: "['segment'] not in index"

In [ ]:
engagement['segment'] = pd.qcut(
    engagement['total_session_time'],
    q=3,
    labels=['Low', 'Medium', 'High']
)

In [ ]:
engagement['segment'].value_counts()

Segmentation Type:
Engagement-Based Segmentation

Methodology:
Customers were segmented based on total session duration. The total session time per customer was calculated and divided into three groups using quantile-based binning (Low, Medium, High).

Segments Created:

High Engagement (402 customers):
Users with the highest total session duration, indicating strong interaction with the platform
Medium Engagement (400 customers):
Users with moderate activity levels
Low Engagement (402 customers):
Users with minimal platform usage

Business Implications
High Engagement:
Likely loyal customers → target for retention and upselling
Medium Engagement:
Opportunity to increase engagement through targeted campaigns
Low Engagement:
At risk → require re-engagement strategies (offers, notifications, onboarding improvements)

In [ ]:
merged_df.head()

In [ ]:
#exporting data into CSV file
merged_df.to_csv("merged_data.csv", index=False)

Why I suggested CSV

Because dataset is in:
Large (~50K rows)
Cleaned in Python
No formatting needed

In [51]:
df_mongo.head()

,_id,timestamp,event_type,page_url,session_duration_sec,browser,os,device_type,timezone,ip_address,referrer,feature,user_id,customer_id
0,69bed815ec39b1e38f11402a,2024-01-16 10:13:48,form_submit,/projects/1779,2263.0,Chrome,iOS 17,mobile,UTC,60.214.112.115,NaN,NaN,410,229
1,69bed815ec39b1e38f11402b,2024-06-12 13:27:31,search,/projects/3627,NaN,Safari,Windows 10,desktop,Asia/Kolkata,NaN,NaN,NaN,2616,14
2,69bed815ec39b1e38f11402c,2025-09-03 13:23:25,notification_click,/api-docs,2915.0,NaN,NaN,NaN,America/Los_Angeles,NaN,NaN,NaN,1292,776
3,69bed815ec39b1e38f11402d,2024-02-19 17:39:16,dashboard_view,/api-docs,730.0,NaN,NaN,NaN,America/Chicago,NaN,NaN,NaN,6066,334
4,69bed815ec39b1e38f11402e,2024-05-05 00:56:31+00:00,import,/team,2353.0,Safari Mobile,macOS 14,desktop,NaN,200.134.219.230,NaN,NaN,3753,115


In [53]:
cursor = conn.cursor()

In [54]:
create_table_query = """
CREATE TABLE IF NOT EXISTS user_activity_cleaned (
    user_id INT,
    customer_id INT,
    event_time DATETIME,
    session_duration_sec FLOAT,
    segment VARCHAR(20),
    company_name VARCHAR(255),
    industry VARCHAR(100),
    is_active BOOLEAN
)
"""

cursor.execute(create_table_query)